### PR TITLE
many: add capability to create missing directories for personal-files interface

### PR DESCRIPTION
**Pull request by ernestl**
**Problem:**

Exploratory code to solve the need to create parent directories in a path specified by the personal files interface when it does not exist. See: https://bugs.launchpad.net/snapd/+bug/1997598

**[Spec](https://docs.google.com/document/d/1yRqiFcdAIUjwakIQdijtW61vKojcJN2vTjrcwJvLAdo/edit)**

**Solution Proposal:**
- Collect "ensure-dir" mount entries from common files interface and add to spec user list and thereby to the user fstab file
- Modify snap-update-ns to action directory creation based on the new ensure-dir mount entries in the user fstab file

**Todo:**
- Unit tests should be fixed and extended once there is more certainty on the direction
- Some open questions and challenges to solve

### CODE REVIEW

File: interfaces/builtin/common_files.go [hunk 1]
```diff
@@ -165,3 +166,65 @@ func (iface *commonFilesInterface) AppArmorConnectedPlug(spec *apparmor.Specific
...
+	if len(pathElements) < 3 {
+		return nil, nil
+	}
+	if pathElements[0] != "$HOME" {
+		return nil, nil
+	}
+
+	pathSpec := &mount.PathSpec{
+		// Path that must exist and will be extended
+		// TODO: Should we expand $HOME?
```
Comment (ernestl) [inline interfaces/builtin/common_files.go:192]: This looks like the place to do dir expansion:
https://github.com/snapcore/snapd/blob/master/cmd/snap-update-ns/user.go#L80-L82

File: cmd/snap-update-ns/update.go [hunk 2]
```diff
@@ -71,7 +76,7 @@ func executeMountProfileUpdate(upCtx MountProfileUpdateContext) error {
 
 	var changesMade []*Change
 	for _, change := range changesNeeded {
-		synthesised, err := change.Perform(as)
+		synthesised, err := change.Perform(upCtx)
```
Comment (ernestl) [inline cmd/snap-update-ns/update.go:79]: Consider adding change to filter out any unknown kinds of desired changes before earlier in executeMountProfileUpdate to make adding new kinds of changes simpler/safer in terms of compatibility if ever needed again

File: cmd/snap-update-ns/change.go [hunk 3]
```diff
@@ -87,7 +88,23 @@ func mimicRequired(err error) (needsMimic bool, path string) {
 	return false, ""
 }
 
-func (c *Change) createPath(path string, pokeHoles bool, as *Assumptions) ([]*Change, error) {
+// TODO: Consider if this can be replaced or become osutil function
+// Created this to ensure links are not followed and error is reported.
+func dirExists(path string) (bool, error) {
```
Comment (mvo5) [inline cmd/snap-update-ns/change.go:93]: We have `osutil.DirExists()` - would that work as well instead of needing another helper?

Comment (ernestl) [reply cmd/snap-update-ns/change.go:93]: Unlike the current osutil version, this version of dirExists() does not follow symlinks, which aligned with the existing snap-update-ns "ensure target" mechanism.

**Pull request closed without merging**